File: client-scalaz/src/main/scala/org/openapitools/client/api/GameResponse.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import GameResponse._

case class GameResponse (
  action: Option[Action],
highestAppVersion: Option[String],
gameId: Option[Long],
gameType: Option[String],
packs: Option[PackListResponse],
scores: Option[ScoreListResponse],
owner: Option[AccountShortResponse],
userPermissionsList: Option[UserPermissionsListResponse],
rules: Option[RuleListResponse],
results: Option[List[NameStringValueResponse]],
ticketsEarned: Option[Long],
title: Option[String],
description: Option[String],
authorOverride: Option[String],
icon: Option[AssetShortResponse],
image: Option[AssetShortResponse],
sequenceType: Option[String],
updatedDate: Option[Long],
startDate: Option[Long],
endDate: Option[Long],
appKey: Option[String])

object GameResponse {
  import DateTimeCodecs._
  sealed trait Action
  case object DATA extends Action
  case object SAVE extends Action
  case object DELETE extends Action
  case object GET extends Action

  object Action {
    def toAction(s: String): Option[Action] = s match {
      case "DATA" => Some(DATA)
      case "SAVE" => Some(SAVE)
      case "DELETE" => Some(DELETE)
      case "GET" => Some(GET)
      case _ => None
    }

    def fromAction(x: Action): String = x match {
      case DATA => "DATA"
      case SAVE => "SAVE"
      case DELETE => "DELETE"
      case GET => "GET"
    }
  }

  implicit val ActionEnumEncoder: EncodeJson[Action] =
    EncodeJson[Action](is => StringEncodeJson(Action.fromAction(is)))

  implicit val ActionEnumDecoder: DecodeJson[Action] =
    DecodeJson.optionDecoder[Action](n => n.string.flatMap(jStr => Action.toAction(jStr)), "Action failed to de-serialize")

  implicit val GameResponseCodecJson: CodecJson[GameResponse] = CodecJson.derive[GameResponse]
  implicit val GameResponseDecoder: EntityDecoder[GameResponse] = jsonOf[GameResponse]
  implicit val GameResponseEncoder: EntityEncoder[GameResponse] = jsonEncoderOf[GameResponse]
}

File: client-clojure/src/sirqul_io_t_platform/specs/game_response.clj
(ns sirqul-io-t-platform.specs.game-response
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.pack-list-response :refer :all]
            [sirqul-io-t-platform.specs.score-list-response :refer :all]
            [sirqul-io-t-platform.specs.account-short-response :refer :all]
            [sirqul-io-t-platform.specs.user-permissions-list-response :refer :all]
            [sirqul-io-t-platform.specs.rule-list-response :refer :all]
            [sirqul-io-t-platform.specs.name-string-value-response :refer :all]
            [sirqul-io-t-platform.specs.asset-short-response :refer :all]
            [sirqul-io-t-platform.specs.asset-short-response :refer :all]
            )
  (:import (java.io File)))


(def game-response-data
  {
   (ds/opt :action) string?
   (ds/opt :highestAppVersion) string?
   (ds/opt :gameId) int?
   (ds/opt :gameType) string?
   (ds/opt :packs) pack-list-response-spec
   (ds/opt :scores) score-list-response-spec
   (ds/opt :owner) account-short-response-spec
   (ds/opt :userPermissionsList) user-permissions-list-response-spec
   (ds/opt :rules) rule-list-response-spec
   (ds/opt :results) (s/coll-of name-string-value-response-spec)
   (ds/opt :ticketsEarned) int?
   (ds/opt :title) string?
   (ds/opt :description) string?
   (ds/opt :authorOverride) string?
   (ds/opt :icon) asset-short-response-spec
   (ds/opt :image) asset-short-response-spec
   (ds/opt :sequenceType) string?
   (ds/opt :updatedDate) int?
   (ds/opt :startDate) int?
   (ds/opt :endDate) int?
   (ds/opt :appKey) string?
   })

(def game-response-spec
  (ds/spec
    {:name ::game-response
     :spec game-response-data}))
